File: xml2json.py
# -*- coding: utf-8 -*-
import xmltodict
import json
from lxml import etree

import xmlutils
import defaultdictvals
import xml2jsonl
import select_by_acc


BIOPROJECT_CONF = {
    "ftp_url": "ftp.ncbi.nlm.nih.gov",
    "ftp_file_path": "bioproject",

}


def xml2json(xml_str, output_path):
    json_doc = xmltodict.parse(xml_str, dict_constructor=dict, force_cdata=True, cdata_key='$')
    return json.dumps(json_doc)


def save_json(json_str, output_path):
    with open(output_path) as f:
        json.dump(json_str, f, indent=2, ensure_ascii=False)


def clear_element(element):
    element.clear()
    while element.getprevious() is not None:
        del element.getparent()[0]


def bioproject_xml_to_dict(file_path:str, output_path:str):
    """
    bioproject.xmlのうち設定された要素をフラットなxmlとしてJSONに変換して書き出す
    :param file_path: bioproject.xmlのパス
    :param output_path: 書き出すJSONファイル名
    :return: void
    """

    # bioproject accのリストによるfilterを追加する場合
    acc = select_by_acc.load_acc()

    context = etree.iterparse(file_path, tag="Package", recover=True)
    dd = defaultdictvals.DefaultDictVal()
    # i = 0
    docs = []
    for events, element in context:
        if element.tag == "Package":
            doc = {}
            doc["type"] = "bioproject"
            doc["identifier"] = dd.get_xattr(element, ".//ProjectID/ArchiveID/@accession")

            # accリストに存在する要素のみ書き出したい場合
            if select_by_acc.select_bp(doc["identifier"], acc):

                doc["organism"] = dd.get_value(element, ".//Project/Project/ProjectDescr/Name")
                doc["title"] = dd.get_value(element, ".//Project/Project/ProjectDescr/Title")
                doc["description"] = dd.get_value(element, ".//Project/ProjectDescr/Description")
                doc["data type"] = dd.get_value(element, ".//ProjectTypeSubmission/ProjectDataTypeSet/DataType")
                doc["organization"] = dd.get_value(element, ".//Submission/Description/Organization/Name")
                doc["publication"] = dd.get_publications(element, ".//Project/ProjectDescr/Publication")
                doc["properties"] = None
                doc["dbXrefs"] = dd.get_xattrs(element, ".//Project/ProjectDescr/LocusTagPrefix/@biosample_id")
                doc["distribution"] = None
                doc["Download"] = None
                doc["status"] = dd.get_value(element, ".//Submission/Description/Access")
                doc["visibility"] = None
                # Todo: 以下ElasticSearchの項目がDate型なため空の値を登録できない（レコードのインポートがエラーとなりスキップされる）
                date_created = dd.get_xattr(element, ".//Submission/@submitted")
                if date_created: doc["dateCreated"] = date_created
                date_mmodified = dd.get_xattr(element, ".//Submission/@last_update")
                if date_mmodified: doc["dateModified"] = date_mmodified
                date_published = dd.get_value(element, ".//Project/ProjectDescr/ProjectReleaseDate")
                if date_published: doc["datePublished"] = date_published

                # そのままxmltodictで機械的にJSONに変換したい場合以下を実行すしリストに追加する
                # ただしxmlが部分的にでもマルフォームであった場合エラーが発生する
                # xml_str = etree.tostring(element)
                # metadata = xml2json(xml_str)
                # docs.append(doc)

                # 特定の要素のみ取り出したい場合
                # if doc["identifier"] == "":
                    #docs.append(doc)

                # jsonlを書き出す場合
                xml2jsonl.dict2jsnl(doc, "bioproject", f"{output_path}l")

        try:
            clear_element(element)
        except TypeError:
            pass

        '''
        if i > 10:
            res = {"bioproject": docs}
            write_json(output_path, docs)
            break
        i += 1
        '''
    # JSONを書き出す場合
    # write_json(output_path, docs)


def bioproject_xml_to_json(file_path: str, output_path:str):
    """
    bioproject.xmlを機械的に階層構造を保ったままJSONに変換して書き出す
    :param file_path:
    :param output_path:
    :return:
    """
    root = xmlutils.read_xml_string(file_path)
    xml2json(root, output_path)


def write_json(out_path:str, docs: dict):
    with open(out_path, "w") as f:
        json.dump(docs, f, indent=4)


if __name__ == "__main__":
    bioproject_xml_to_dict("bioproject.xml", "bioproject.json")

File: rbversion/xml2json.rb
require 'json'
require 'ox'

#
# bioproject.xmlをjsonに変換するクラス
#
class BPXml2JsonConverter
    #
    # コンストラクタ
    #
    # @param [String] bp_xml bioproject.xml から切り出したXMLエレメント
    #
    def initialize(bp_xml)
        begin
            xml = Ox.parse(bp_xml)
            project = xml.Project
        rescue
            print bp_xml
            raise
        end

        @hash = {
            'type' => 'bioproject',
            'identifier' => project.locate('Project/ProjectID/ArchiveID/@accession')[0], # "//ProjectID/ArchiveID/@accession"
            'organism' => project.locate('Project/ProjectDescr/Name')[0]&.text, # "//Project/Project/ProjectDescr/Name"
            'title' => project.locate('Project/ProjectDescr/Title')[0]&.text, # "//Project/Project/ProjectDescr/Title"
            'description' => project.locate('Project/ProjectDescr/Description')[0]&.text, # "//Project/ProjectDescr/Description"
            'data type' =>  project.locate('Project/ProjectTypeSubmission/ProjectDataTypeSet/DataType')[0]&.text, # "//ProjectTypeSubmission/ProjectDataTypeSet/DataType"
            'organization' => project.locate('Project/Submission/Description/Organization/Name')[0]&.text, # "//Submission/Description/Organization/Name"
            'publication' => project.locate('Project/ProjectDescr/Publication').map do |elm| # "//Project/ProjectDescr/Publication"
                {'id' => elm.id, 'Title' => elm.locate('Title')[0]&.text}
            end,
            'properties' => nil,
            'dbXrefs' => project.locate('Project/ProjectDescr/LocusTagPrefix').map do |elm| # "//Project/ProjectDescr/LocusTagPrefix/@biosample_id"
                elm.locate('@biosample_id')
            end,
            'distribution' => nil,
            'Download' => nil,
            'status' => project.locate('Project/Submission/Description/Access')[0]&.text,
            'visibility' => nil
        }
        # 以下、Pythonコードからコピー
        # Todo: 以下ElasticSearchの項目がDate型なため空の値を登録できない（レコードのインポートがエラーとなりスキップされる）
        submission = project.locate('Project/Submission')[0]
        if submission
            @hash['dateCreated'] = submission.submitted if submission.submitted
            @hash['dateModified'] = submission.last_update if submission.last_update
        end
        date_published = project.locate('Project/ProjectDescr/ProjectReleaseDate')[0]&.text
        @hash['datePublished'] = date_published if date_published
    end

    #
    # XMLエレメントを所定の形式のjsonに変換します。
    #
    def to_json
        @hash.to_json
    end

    #
    # XMLエレメントから所定のメタ情報のjsonを出力します。
    #
    def meta_json
        {index: {_index: 'bioproject', _type: 'metadata', _id: @hash['identifier']}}.to_json
    end

    #
    # XMLエレメントをメタ情報と所定の形式のjsonのセットで出力します。
    #
    def to_json_with_meta
        "#{meta_json}\n#{to_json}"
    end
end

#
# Python の etree.iterparse と類似の機能を提供するクラス
#
class XmlFileLazyReader
    #
    # コンストラクタ
    #
    # @param [String] file_path XMLファイルのパス
    # @param [String] tag XMLファイルから分割して切り出すタグ
    # @param [Integer] buffer_kbytes XMLファイルを読み込む際のバッファーサイズ(KB)。初期値: 16
    #
    def initialize(file_path, tag, buffer_kbytes = 16)
        raise ArgumentError, 'file_path is not set.' unless file_path.is_a?(String)
        raise ArgumentError, 'tag is not set.' unless tag.is_a?(String)
        raise ArgumentError, 'buffer_kbytes is not Integer.' unless buffer_kbytes.is_a?(Integer)

        @file = File.open(file_path, 'r')
        @buf = ''
        @buffer_size = buffer_kbytes * 1024
        @fiber = Fiber.new do
            @buf = @file.read(@buffer_size)

            loop do
                res = find_tag_index(tag, 0)
                break unless res
                Fiber.yield(@buf.slice!(0..res[:stop_index])[res[:start_index]..-1])
            end
        end
    end

    #
    # XMLを分割したタグごとに取得します。
    #
    def each_element
        loop do
            yield(@fiber.resume)
            unless @fiber.alive?
                return
            end
        end
    end

    private
    def find_tag_index(tag, start_index)
        current_index = start_index
        index1 = 0
        loop do
            index1 = @buf.index("<#{tag}>", current_index)
            while index1 == nil
                return nil unless bufnext
            end
            index2 = @buf.index("<#{tag}>", index1 + tag.length + 1)
            index3 = @buf.index("</#{tag}>", index1 + tag.length + 1)

            # while index3 == nil # 同一タグのネストを考慮 *する* 場合
            while index3 == nil && index2 == nil # 同一タグのネストを考慮 *しない* 場合
                return nil unless buffer_next
                index3 = @buf.index("</#{tag}>", @buffer_size - tag.length - 3)
            end

            # 同一タグのネストを考慮 *する* 場合の実装
            # if index2 && index2 < index3
            #     res = find_tag_index(tag, index2)
            #     current_index = res[:stop_index] + 1
            # else
            #     return {start_index: index1, stop_index: index3 + tag.length + 2}
            # end

            # 同一タグのネストを考慮 *しない* 場合の実装
            if index2 && !index3
                print 'close tag lost '
                index3 = index2 - 1
            end
            return {start_index: index1, stop_index: index3 + tag.length + 2}

        end while(index1)
        return nil
    end

    def buffer_next
        tmp = @file.read(@buffer_size)
        return nil unless tmp
        if @buf.length > 100 * 1024 * 1024
            print "Element size over 100MB [#{@buf.slice(@buf.index('<ArchiveID'), 80)}]"
            @buf = ''
        end
        @buf += tmp
    end
end

#
# bioproject.xml を BPXml2JsonConverter によって jsonl 形式で出力するためのクラス
#
class BPXml2Jsonl
    #
    # コンストラクタ
    #
    # @param [String] out_path jsonファイルの出力先パス
    #
    def initialize(out_path)
        raise ArgumentError, 'out_path is not set.' unless out_path.is_a?(String)

        @to_json_ractor = Ractor.new(out_path) do |out_path|
            File.delete(out_path)
            json = File.open(out_path, 'a')
            loop do
                bp_xml = Ractor.receive
                break unless bp_xml
                json << BPXml2JsonConverter.new(bp_xml).to_json_with_meta + "\n"
            end
        end
    end

    #
    # 渡されたXMLエレメントを変換処理を行うスレッドに渡します。
    #
    # @param [String] xml 変換対象のXMLエレメント
    #
    def send(xml)
        @to_json_ractor.send(xml)
    end
end

print Time.now, " Started\n"
reader = XmlFileLazyReader.new('C:\Users\iota_\Downloads\bioproject.xml', 'Package', 32768)
xml2json = BPXml2Jsonl.new('bioproject.json')
reader.each_element do |elm|
    xml2json.send(elm)
end
print Time.now, " Finished\n"
